### ISSUE TITLE
Some unhashable types are not detected as unhashable

### ISSUE THREAD

**Issue by TheTripleV** [microsoft/pyright#5446]
**Describe the bug**
```python
class A:
    __hash__ = None # type: None

_ = {A(): 1}
```
correctly reports `Dictionary key must be hashable\n Type "A" is not hashable`. (Unrelated - The `# type: None` is required, which I'm not a fan of but I can leave that for a future issue).
I believe this error message was added in #4145.

However,

```python
# To Reproduce
class B:
    def __eq__(self, value):
        ...

_ = {B(): 1}
````
does not result in a type error even though it is invalid and results in a runtime error. `B` is unhashable according to [the python docs](https://docs.python.org/3/reference/datamodel.html#object.__hash__:~:text=if%20it%20defines%20__eq__()%20but%20not%20__hash__()%2C%20its%20instances%20will%20not%20be%20usable%20as%20items%20in%20hashable%20collections)

This was covered in #3223 and was marked `as designed` but that was before support for unhashable type errors was added to pyright.

**Expected behavior**
The line `_ = {B(): 1}` should result in `Dictionary key must be hashable\n Type "B" is not hashable`

**VS Code extension or command-line**
- Python 3.11.1, Windows 11 x64
- Running via VSCode Extension
- Pylance language server 2023.7.11 (pyright 74d8f3c4 (tag 1.1.316))

**Comment by TheTripleV** [comment]
I took a stab at this but ran into issues because the builtins in typeshed define eq but not hash, making them unhashable according to the hints but are hashable in reality. This causes a ton of false positives.

**Comment by erictraut** [comment]
Pyright has no way of knowing whether a class properly implements `__hash__` other than to use the type information it is given. The convention established in typeshed is that all classes are assumed to be hashable (by virtue of `object` defining a `__hash__` method). Classes that are not hashable must explicitly redefine hash as `__hash__: None = None`. Pyright honors this convention.

If you would like to bring this up with the maintainers of typeshed, please feel free to do so.

If this were a false positive, I'd feel more strongly about addressing it, but it's a false negative. There are many runtime errors that a type checker cannot detect. AFIAK, mypy doesn't attempt to check for hashability under any circumstance, so pyright is already going beyond what other type checkers attempt to do here.

In the meantime, I don't think there's anything pyright can do to eliminate this edge-case false negative. Doing so would generate many false positives in every typed code base.

**Comment by TheTripleV** [comment]
Some fixes were made to typeshed. I took another go afterwards and it seems to work without false positives.
The additional triggering cases I added are
1. If a class does not define hash and defines eq, it's unhashable.
2. If a class does not define hash or eq, and all base classes are not hashable, then the type is unhashable.

2 errs on the side of false negatives. It would be more accurate to look at the order of classes in the mro.

I'll rebase the linked draft pr the next time pyright pulls typeshed.

**Comment by erictraut** [comment]
I'm uncomfortable with the proposed rules. Examining base classes does not look sound to me. I would expect any rule to look at the class hierarchy and honor the MRO when determining whether any particular methods are implemented.

How does the CPython code determine whether a class is hashable? What is the exact logic it uses? If we're going to fix this (and I admit that I'm still not convinced it's important enough to fix), then we should be consistent with the actual runtime rather than creating heuristics that attempt to approximate it. In my experience, such heuristics tend to cause many problems.

**Comment by btidor** [comment]
Hi @erictraut, I just ran into this issue as well. Here's my understanding of the problem:

* Python determines hashability as you would expect: if a class defines `__hash__` (and it's not `None`), then the class is hashable: https://github.com/python/cpython/blob/3.12/Lib/_collections_abc.py#L124-L128.

* The problem is that Python has a special case where overriding `__eq__` automatically causes `__hash__` to be overridden and set to `None`. As I understand it, Pyright doesn't currently model this behavior.

From the [docs](https://docs.python.org/3/reference/datamodel.html#object.__hash__):

> A class that overrides `__eq__()` and does not define `__hash__()` will have its `__hash__()` implicitly set to None. When the `__hash__()` method of a class is None, instances of the class will raise an appropriate TypeError when a program attempts to retrieve their hash value, and will also be correctly identified as unhashable when checking isinstance(obj, collections.abc.Hashable).

Here's a minimal example. You can see that `B.__hash__` is `None` even though we didn't override it explictly:

```py
from collections.abc import Hashable
from typing import Any

class A:
    pass

class B:
    def __eq__(self, other: Any) -> bool:
        raise NotImplementedError

a = A()
assert isinstance(a, Hashable)
assert "__eq__" not in A.__dict__
assert "__hash__" not in A.__dict__

b = B()
assert not isinstance(b, Hashable)
assert B.__dict__["__eq__"] is B.__eq__
assert B.__dict__["__hash__"] is None
```

If we add this condition at the end, Pyright reports that it always evaluates to False. At runtime, the condition always evaluates to True:

```py
if B.__hash__ is None:
    ...
```

Thanks!

**Comment by erictraut** [comment]
I've read the above documentation, but it omits some important details. In particular, it doesn't explain _when_ or _how_ the CPython code implicitly sets `__hash__` to `None`. For example, can a class implement an `__eq__` method but then use a mix-in class to implement a `__hash__` method?

The cpython code that you linked to above doesn't do anything with the `__eq__` method, so I presume there is more to it than that.

**Comment by hmc-cs-mdrissi** [comment]
I think this is main place, https://github.com/python/cpython/blob/4b4a5b70aa8d47b1e2a0582b741c31b786da762a/Objects/typeobject.c#L6794, + tp_inheritance logic in that file for how it’s determined whether hash is defined on a type or not.

**Comment by erictraut** [comment]
Reading the CPython sources further, it appears to base its decision on the "dominant base" in the case where there are multiple base classes. I presume that's the first base class in the list. That means mix-in classes shouldn't affect the logic. It also means that the mro should not be used for determining hashability, since only the chain of "dominant bases" is applicable.

@TheTripleV, I'm going to close your PR because it doesn't match the CPython implementation. I'll reopen this issue as a feature request.

**Comment by TheTripleV** [comment]
Yes, I just tested hashability on some permutations of superclasses. Mixins (like those in collections.abc) don't affect things in a special way. Going by mro order, the decision is based the on the first class that is explicitly unhashable (defines eq or sets hash to none) or explicitly hashable (defines hash). Hashable otherwise. I assume this is what you mean by "dominant class"

**Comment by erictraut** [comment]
This will be included in the next release. Thanks to @TheTripleV for the PR.

**Comment by erictraut** [comment]
Based on the discussion in https://github.com/microsoft/pyright/issues/5513, I'm going to revert this change for now. We can revisit it in the future if/when typeshed is updated accordingly.

As I mentioned above, I consider this a nice-to-have feature, but it's not worth it if it introduces a bunch of false positive errors.
